### ISSUE TITLE
LaTeX parsing based on LLM (Large language model)

### ISSUE THREAD

**Issue by 8vasu** [sympy/sympy#26128]
A discussion on developing an LLM-based LaTeX parser was initiated here: https://github.com/sympy/sympy/pull/26098. That discussion is being moved here.

[List of local LLMs](https://python.langchain.com/docs/guides/local_llms)

---

## List of existing parsers

- **TeX Live / pdflatex**
  - [TeX Live](https://www.tug.org/texlive/)
  - Language: WEB (Pascal and C)
  - Output formats: PDF, DVI

- **MiKTeX**
  - [MiKTeX](https://miktex.org/)
  - Language: C++
  - Output formats: PDF, DVI

- **XeTeX / xelatex**
  - [XeTeX](http://xetex.sourceforge.net/)
  - Language: WEB (Pascal and C), C++
  - Output formats: PDF, DVI, (XDV with intermediate format)

- **LuaTeX / lualatex**
  - [LuaTeX](http://www.luatex.org/)
  - Language: WEB (Pascal and C), Lua
  - Output formats: PDF, DVI

- **MathJax**
  - [MathJax](https://www.mathjax.org/)
  - Language: JavaScript
  - Output formats: HTML with CSS, MathML, SVG

- **KaTeX**
  - [KaTeX](https://katex.org/)
  - Language: JavaScript
  - Output formats: HTML with CSS

- **LaTeX.js**
  - [LaTeX.js](https://latex.js.org/)
  - Language: JavaScript
  - Output formats: HTML

- **Pandoc**
  - [Pandoc](https://pandoc.org/)
  - Language: Haskell
  - Output formats: HTML, PDF, DOCX, EPUB, and many others (with appropriate output converters)

- **Tectonic**
  - [Tectonic](https://tectonic-typesetting.github.io/en-US/)
  - Language: Rust
  - Output formats: PDF

- **LyX**
  - [LyX](https://www.lyx.org/)
  - Language: C++
  - Output formats: PDF, HTML, RTF, and others via export

- **MathType**
  - [MathType](https://www.wiris.com/en/mathtype/)
  - Language: Unknown (proprietary software)
  - Output formats: HTML, MathML, LaTeX

- **Overleaf**
  - [Overleaf](https://www.overleaf.com/)
  - Language: Primarily JavaScript (online platform)
  - Output formats: PDF (online LaTeX editor)

- **LaTeXML**
  - [LaTeXML](https://github.com/brucemiller/LaTeXML)
  - Language: Perl
  - Output formats: XML, HTML, ePub, MathML

- **pylatexenc**
  - [pylatexenc](https://github.com/phfaist/pylatexenc)
  - Language: Python
  - Output formats: unicode

- **latex2sympy (various incarnations)**
  - [augustt198](https://github.com/augustt198/latex2sympy), [grasple](https://github.com/grasple/latex2sympy), [purdue-tlt](https://github.com/purdue-tlt/latex2sympy), (OrangeX4)[https://github.com/OrangeX4/latex2sympy], (PiCalc)[https://github.com/PiCalc/latex2sympy], (jackatbancast)[https://github.com/jackatbancast/latex2sympy], etc.
  - Language: Python
  - Output formats: SymPy

- **SymPy (latex2sympy-derived ANTLR-based backend)**
  - [Sympy](https://docs.sympy.org/latest/modules/parsing.html#mathrm-latex-parsing-functions-reference)
  - Language: Python, ANTLR
  - Output formats: SymPy, LaTeX, MathML, pretty print (ASCII/Unicode)

- **SymPy (Lark-based backend)**
  - [Sympy](https://docs.sympy.org/latest/modules/parsing.html#mathrm-latex-parsing-functions-reference)
  - Language: Python, Lark
  - Output formats: SymPy, LaTeX, MathML, pretty print (ASCII/Unicode)

- **Matplotlib's mathtext**
  - [matplotlib](https://matplotlib.org/)
  - Language: Python
  - Output formats: Raster images (PNG), Vector images (SVG), and others via matplotlib's backends

- **Emacs Calc**
  - [Emacs Calc](https://www.gnu.org/software/emacs/manual/html_mono/calc.html)
  - Language: Emacs Lisp
  - Output formats: Can output to [SageMath](https://www.sagemath.org/)  via a combination of `math-read-expr` and `math-compose-expr`. See [symtex.el](https://github.com/ultronozm/symtex.el/blob/calc/symtex.el) for an example.

**Comment by 8vasu** [comment]
@asmeurer @wermos @sylee957

**Comment by sylee957** [comment]
- Is locally trained model is a direction we should take, or is prompt engineering a direction we should take? Or otherwise is there other direction we can use.
- Is there any previous success on this?
- How much resources we need. Essentially should we use funding for computational powers. 
- How much influence does it have for users? If we are asking other people to invest time on this

**Comment by 8vasu** [comment]
To make the process more democratic, I think we should vote after considering the above table of pros and cons.

To vote for *local LLMs*, please react with a :+1: 

To vote for *cloud-based LLMs*, please react with a :heart:

**Comment by 8vasu** [comment]
I voted for local LLMs because of the non-requirement of internet connectivity and because I have not worked with them before and feel adventurous (although I am note sure if my personal computer has enough computational power to handle them; maybe experts can help me with this?)

Maybe after we have determined if we will use local or cloud-based LLMs, we can create another list of Pros and Cons for the various local or cloud-based options available and have another poll.

Note: if the maintainers think that conducting polls for this is not the correct way, then I will cancel the above comment.

**Comment by sylee957** [comment]
I’d vote for cloud based LLM, especially for the exploration stage, I think that we don’t even know how to use LLM properly at all, before we investigating any expert options like reducing cost, fine tune.
I may take a chance to follow some definite learning resources.

**Comment by asmeurer** [comment]
I'm personally a little curious to see if fine-tuning could improve model output here. But one issue is that we don't have a good dataset to train on. We could train one by synthetically generating expressions and using `latex()` on them, but that would only generate a subset of possible latex. We could also extract a lot of formulas from somewhere like arxiv papers, and use the current parser on them, but it would be hard to judge the accuracy of it without manual checking. Plus ideally we'd like to support things that aren't currently supported by the current parser. 

Outside of fine-tuning, I don't think there's a big difference between online and local models. The main thing that would need to be done is construct a good prompt for the problem. Different models will have different levels of capability.

**Comment by 8vasu** [comment]
>  it would be hard to judge the accuracy of it without manual checking.

The only thing we can completely trust without manual checking is some actual engine such as TeX, eTeX, pdfTeX, XeTeX, LuaTeX, etc. Consider the following 2 LaTeX expressions:

1. `[\begin{matrix}...\end{matrix}]`
2. `[  {  \begin{matrix}...\end{matrix}   }   ]`

In https://github.com/sympy/sympy/pull/26098, I wrote some parser rules which treat the above two equally. Latex engines like pdfTeX also produce identical output with those inputs, so maybe they first convert the input to some intermediate format which is more "definite" in the sense that it defines a (mathematical) function with the following properties:

1. The inputs of the function are LaTeX math expressions and outputs are in this hypothetical "intermediate format".
2. The function is many-to-one. For instance, the above two LaTeX `matrix` expressions produce the same "sanitized" output in this hypothetical format.
3. If we are lucky, then the range of this function is much smaller than the domain, i.e. our "intermediate format to SymPy translator" (LLM or otherwise) will need to focus on far fewer inputs than if the input was in LaTeX.

I know [Lars](https://ctan.org/author/madsen) ["daleif"](https://github.com/daleif) Madsen personally; he might know about this; I will ask him soon.

**Comment by asmeurer** [comment]
This is related but a bit orthogonal to the LLM idea, but I would love to have a LaTeX math-mode parser library for Python. Unlike the SymPy parser, it should just parse to an AST or CST, and support any valid math-mode LaTeX. 

In particular what I want to do is build a LaTeX math autoformatter that does things like remove redundant `{}`, remove unnecessary `\left` and `\right`, and normalize spaces for readability. That could then be used as an autoformatter on the math in our docs, and also be used as a post-processor on the `latex()` function. It could also be used to build a LaTeX math linter that warns about common LaTeX mistakes like `sin(x)` (instead of `\sin(x)`) or `x^12`. This is on my list of projects that I would love have but haven't had enough time to build.

**Comment by 8vasu** [comment]
I talked with Lars Madsen about this today, and he agrees with me in that the only way to do this that guarantees complete reliability is to use some actual TeX engine (TeX, eTeX, pdfTeX, XeTeX, LuaTeX, etc).

His suggestion was to email the LuaTeX folks to ask if there exists any Lua hooks that allow programmers to access some intermediate format (maybe an AST or CST). I will email them this weekend.

Lua is extremely easy to embed anywhere; for example: https://labix.org/lunatic-python.

**Comment by asmeurer** [comment]
The Python bit isn't critical. I would mostly want that so that it is easier for people to contribute to. And also because it's what I know :)

**Comment by asmeurer** [comment]
> Also unrelated, but adding this here since Lars Madsen mentioned this: [typst/typst](https://github.com/typst/typst).

Yep, we are familiar with typst. https://github.com/sympy/sympy/issues/24959. Maybe one day we can use it in our docs, if Sphinx etc. get good support for it. But for the SymPy printer/parser we have to follow what people are using and right now that's LaTeX. (also an LLM is going to be much better at parsing LaTeX because it's seen a ton of LaTeX in its training data)

**Comment by 8vasu** [comment]
> But for the SymPy printer/parser we have to follow what people are using and right now that's LaTeX.

Of course. I do not see the mathematics community moving away from LaTeX anytime soon.

**Comment by 8vasu** [comment]
I have not emailed LuaTeX yet, but I did some quick research myself.

## Nodes

According to GPT4 API:

> TeX converts the input into an intermediate format consisting of lists of nodes. These nodes represent boxes, glue (space), penalties, and other elements that TeX uses to construct the final output. This happens after TeX has parsed and processed the input but before the final typesetting occurs.

This is an introductory example: https://wiki.luatex.org/index.php/Traversing_TeX_nodes. Notice how the unicode ε is detected as char=941.

This is a link to the list of nodes: https://mirrors.ibiblio.org/CTAN/systems/doc/luatex/luatex.pdf#page=132. This list is properly organized; for example, math-specific nodes are separately listed.

## Callbacks

According to GPT4 API:

> In LuaTeX, **callbacks** are hooks that allow you to register multiple functions to be called at specific points in the processing of the document. When a particular event occurs, LuaTeX will execute all the functions that have been registered to that callback, in the order they were added.

One can add their own Lua functions to a callback using `luatexbase.add_to_callback`.

Here is a list of callbacks: https://wiki.luatex.org/index.php/Callbacks and a more condensed (but unverified) list of callbacks produced by GPT4:

- `start_run`: Called at the start of the TeX run.
- `stop_run`: Called at the end of the TeX run.
- `start_page_number`: Called when the page number is incremented.
- `stop_page_number`: Called after the page number is incremented.
- `pre_dump`: Called before the format file is dumped.
- `start_file`: Called when a file is opened.
- `stop_file`: Called when a file is closed.
- `show_error_hook`: Called when an error is shown.
- `show_warning_hook`: Called when a warning is shown.
- `show_info_hook`: Called when an info message is shown.
- `process_input_buffer`: Called before TeX processes a line of input.
- `process_output_buffer`: Called before data is written to an output file.
- `process_jobname`: Called when TeX is determining the job name.
- `open_read_file`: Called when a file is opened for reading.
- `read_font_file`: Called when a font file is read.
- `find_write_file`: Called when a file is opened for writing.
- `find_output_file`: Called when the output file is opened.
- `find_format_file`: Called when the format file is opened.
- `find_read_file`: Called when a file is opened for reading.
- `find_font_file`: Called when a font file is searched.
- `find_vf_file`: Called when a virtual font file is searched.
- `find_map_file`: Called when a font map file is searched.
- `find_enc_file`: Called when an encoding file is searched.
- `find_sfd_file`: Called when a subfont definition file is searched.
- `find_opentype_file`: Called when an OpenType font file is searched.
- `find_truetype_file`: Called when a TrueType font file is searched.
- `find_type1_file`: Called when a Type1 font file is searched.
- `find_data_file`: Called when a data file is searched.
- `find_cidmap_file`: Called when a CID-to-GID mapping file is searched.
- `find_pk_file`: Called when a PK font file is searched.
- `find_image_file`: Called when an image file is searched.
- `find_ocp_file`: Called when an OCP file is searched.
- `find_ocf_file`: Called when an OCF file is searched.
- `find_cmap_file`: Called when a CMap file is searched.
- `find_fontmap_file`: Called when a font mapping file is searched.
- `read_vf_file`: Called when a virtual font file is read.
- `read_map_file`: Called when a font map file is read.
- `read_enc_file`: Called when an encoding file is read.
- `read_cidmap_file`: Called when a CID-to-GID mapping file is read.
- `read_pk_file`: Called when a PK font file is read.
- `read_ocp_file`: Called when an OCP file is read.
- `read_ocf_file`: Called when an OCF file is read.
- `read_cmap_file`: Called when a CMap file is read.
- `read_fontmap_file`: Called when a font mapping file is read.
- `define_font`: Called when a font is defined.
- `hyphenate`: Called for hyphenation of a list of nodes.
- `ligaturing`: Called for ligature creation in a list of nodes.
- `kerning`: Called for kerning in a list of nodes.
- `pre_linebreak_filter`: Called before the line-breaking algorithm starts.
- `buildpage_filter`: Called when TeX decides whether material is to be added to the current page or held over for the next page.
- `hpack_filter`: Called when a horizontal box is being packed.
- `vpack_filter`: Called when a vertical box is being packed.
- `pre_output_filter`: Called before the output routine starts.
- `linebreak_filter`: Called when the line-breaking algorithm is applied to a paragraph.
- `append_to_vlist_filter`: Called when material is added to the main vertical list.
- `mlist_to_hlist_filter`: Called when a math list is converted to a horizontal list.
- `finish_pdffile`: Called when the PDF file is finished.
- `finish_pdfpage`: Called when a PDF page is finished.
- `pre_pdfpage_filter`: Called before a PDF page is shipped out.
- `post_pdfpage_filter`: Called after a PDF page is shipped out.
- `contribute_filter`: Called when material is contributed to a list.
- `call_edit`: Called when the `--edit` option is used.

## Example

I asked GPT4 API to combine nodes and callbacks to parse a matrix. I have not tested this, but I am pasting it here anyway:

> To parse a LaTeX matrix using LuaTeX nodes and callbacks, you can use the luacode environment to write Lua code within your LaTeX document. Below is an example of how you might set up a callback to parse a matrix environment and manipulate its nodes using LuaTeX.

```latex
\documentclass{article}
\usepackage{luacode}
\usepackage{amsmath}

\begin{luacode*}
function parse_matrix(head)
  for n in node.traverse_id(node.id("hlist"), head) do
    -- Assuming the matrix is an hlist, you can iterate over its elements
    for current_node in node.traverse(n.head) do
      -- Do something with each element of the matrix
      -- For example, you could print the node type and content if it's a character
      if current_node.id == node.id("glyph") then
        texio.write_nl("Character node: " .. unicode.utf8.char(current_node.char))
      end
    end
  end
  return true
end

luatexbase.add_to_callback("hpack_filter", parse_matrix, "parse_matrix")
\end{luacode*}

\begin{document}

\[
\begin{matrix}
  a & b \\
  c & d \\
\end{matrix}
\]

\end{document}
```

GPT4 API:

> In this example, the parse_matrix function is registered as an hpack_filter callback, which is called whenever an hlist is packed. The function iterates over the nodes in the hlist, and if it encounters a glyph node (which represents a character), it prints the character to the log.
>
> Please note that this is a simplified example and actual matrix parsing would require a more complex approach to correctly identify and manipulate the matrix structure. The LaTeX matrix environment can be deeply nested and may contain different types of nodes that need to be handled appropriately.

**Comment by 8vasu** [comment]
If I was to summarize the above comment in one sentence and in very simple terms, I think that the "Nodes" are somewhat like "tokens" in Lark, so that part is done for us by LuaTeX;  one only needs to write their own analog of `transformer.py` (in Lua or Python or whatever) or feed the nodes into an LLM.

In any case, I might get a little time during summer to think about this approach, but maybe first we should decide whether a local or a cloud-based LLM should be used. There is no rush at all, but just wanted to mention that maybe the maintainers could make the decision OR invite people to vote here: https://github.com/sympy/sympy/issues/26128#issuecomment-1914786705.

**Comment by Upabjojr** [comment]
> We could train one by synthetically generating expressions and using `latex()` on them, but that would only generate a subset of possible latex.

Can we use a VLM to generate LaTeX out of PDFs?

**Comment by wermos** [comment]
> This is related but a bit orthogonal to the LLM idea, but I would love to have a LaTeX math-mode parser library for Python. Unlike the SymPy parser, it should just parse to an AST or CST, and support any valid math-mode LaTeX.

I came to this issue to actually discuss this thing itself. I recently came across [`tree-sitter`](https://tree-sitter.github.io/tree-sitter/), which is a parsing library, with existing grammars for many languages, including LaTeX. `tree-sitter` has [Python bindings](https://github.com/tree-sitter/py-tree-sitter/). The LaTeX grammar for `tree-sitter` is [here](https://github.com/latex-lsp/tree-sitter-latex). This LaTeX grammar is used in [LSP](https://microsoft.github.io/language-server-protocol/)s, so we know that it is reasonably powerful.

The issue that I had, while working on this project, was the dual focus on making a grammar for LaTeX as well as having it work with SymPy.

If we can offload writing the grammar to an external project, then the problem simply becomes of understanding their grammar, and then walking the tree and constructing a SymPy expression as we go.

**Comment by asmeurer** [comment]
tree-sitter indeed sounds like a good place to start for that. How much does it parse inside of math expressions?

**Comment by halfrot** [comment]
> > We could train one by synthetically generating expressions and using `latex()` on them, but that would only generate a subset of possible latex.
> 
> Can we use a VLM to generate LaTeX out of PDFs?

Seems to be pretty easy for current VLM or OCR models like Qwen-VL or deepseek-OCR :)
